src/EditorFeatures/Core/EditAndContinue/PdbMatchingSourceTextProvider.cs

```diff
@@ -0,0 +1,173 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.EditAndContinue
+{
+    /// <summary>
+    /// Notifies EnC service of host workspace events.
+    /// </summary>
+    [ExportEventListener(WellKnownEventListeners.Workspace, WorkspaceKind.Host), Shared]
+    [Export(typeof(PdbMatchingSourceTextProvider))]
+    internal sealed class PdbMatchingSourceTextProvider : IEventListener<object>, IEventListenerStoppable, IPdbMatchingSourceTextProvider
+    {
+        private readonly object _guard = new();
+
+        private bool _isActive;
+        private int _baselineSolutionVersion;
+        private readonly Dictionary<string, (DocumentState state, int solutionVersion)> _documentsWithChangedLoaderByPath = new();
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public PdbMatchingSourceTextProvider()
+        {
+        }
+
+        public void StartListening(Workspace workspace, object serviceOpt)
+        {
+            workspace.WorkspaceChanged += WorkspaceChanged;
+        }
+
+        public void StopListening(Workspace workspace)
+        {
+            workspace.WorkspaceChanged -= WorkspaceChanged;
+        }
+
+        private void WorkspaceChanged(object? sender, WorkspaceChangeEventArgs e)
+        {
+            if (!_isActive)
```

Comment:
Does this need to use the guard?

Reply:
I don't think so - this is just quick way out, we check again when we take the lock.